### PR TITLE
Add text modality, and some basic refactoring

### PR DESCRIPTION
**Pull request by bhavul**
PR to add text modality and text based training to our gato model here.

Changes include: 
- Introducing huggingface datasets dependency since we can mostly find our datasets from there
- Some refactoring in code to distinguish between control tasks and text tasks, as well as modifying those parts of the code to be more extendable
- GPT2 Text Tokenizer, and next token prediction 
- Adding perplexity metrics (although I should note that original Gato paper does not showcase any evaluation metrics of texts)
- Some variable renames to make their names more clearer on what they represent / purpose they serve

The code has been written in a way that it follows similar conventions to current codebase, and its structure. 

**Dataset (caveat)**
We ultimately wish to test it on [pile dataset](https://huggingface.co/datasets/EleutherAI/pile), however for testing sake, I found a smaller sample [pile-10k](https://huggingface.co/datasets/NeelNanda/pile-10k). I tried this but it throws some error mostly meaning it's not maintained properly on huggingface hub. We could potentially also try [lighteval/pile](https://huggingface.co/datasets/lighteval/pile).  
This is why the default is [wikitext](https://huggingface.co/datasets/wikitext), which is a reliable dummy text dataset.

**Testing**
To test the code, I installed all the packages listed in env.yml locally in a conda environment (except for CUDA since I was on a CPU machine), and ran this : 
```
python train.py --embed_dim=768 --layers=6 --heads=24 --training_steps=1000 --log_eval_freq=10 --warmup_steps=10 --batch_size=4 -k=240 --eval_episodes=10 --sequence_length=1024 --activation_fn=gelu --save_model --save_mode=checkpoint --text_prop 1.0 --cpu
```

**References**
For more on how we can make changes to the repo, or understand its different parts - https://docs.google.com/document/d/16rNVqRucDMz4yqiYJz8NZFPHMGvzO0emb0KkxjAlfXo/edit#heading=h.6kliytos981a

### CODE REVIEW

File: gato/training/trainer.py [hunk 1]
```diff
@@ -105,9 +108,13 @@ def train_iteration(self, num_steps, iter):
     def train_step(self):
         logs = {}
         logs['training/learning_rate'] = self.scheduler.get_lr()[0] # store LR at current step
-        
+
         # Build training batch
-        batch_dicts = self.sample_control_batch(self.args.batch_size)
+        # for simplicity sake - assume interleave training on control & text
+        if self.steps % 2 == 0:
```
Comment (daniellawson9999) [inline gato/training/trainer.py]: I think we should try to implementing mixing both text/control in one batch. Maybe have a configurable argument like text_prop(value from 0 to 1).  Sample texst_prop * batch_size sequences of text, say control_prop = 1-text_prop for now, and control batch size is control_prop * batch_size. Sample sub batches of these sizes, join them together into one batch, and take a gradient up using this combined batch. We may want to restructure tokenize_input_dicts and forward() to potentially make this easier, can elaborate on a separate comment.

File: train.py [hunk 2]
```diff
@@ -27,11 +28,14 @@ def main(args):
 
     envs, datasets = load_envs(args.datasets) # Load Minari datasets and corresponding Gym environments
 
+    # load the text dataset
+    text_dataset = load_dataset('wikitext', 'wikitext-103-raw-v1')
```
Comment (daniellawson9999) [inline train.py]: let's make this a bit more configurable. Potentially follow pattern in args.datasets (where one can pass a list of strings with the names of the control datasets), only load names of datasets specified in args, load none by default

File: gato/training/trainer.py [hunk 3]
```diff
@@ -116,13 +123,31 @@ def train_step(self):
...
+        text_tokens = [self.model.text_tokenizer.encode(text) for text in selected_text_examples]
+        input_dict['text'] = torch.stack(text_tokens).to(self.args.device)
+        # The targets are the same as the input, but shifted one position to the left
+        # append the appropriate end-of-sequence token as well
+        input_dict['text_targets'] = torch.cat(
+            [tokens[1:], torch.tensor([self.model.text_tokenizer.tokenizer.eos_token_id]).expand(1, tokens.shape[1])],
+            dim=0
+        ).to(self.args.device)
+
+        return input_dict
```
Comment (daniellawson9999) [inline gato/training/trainer.py]: Text is already tokenized (text_tokenizer.encode), right now, as implemented in tokenize_input_dicts(), it would try to tokenize a second time.

File: train.py (hunk 2 above)
Comment (daniellawson9999) [inline train.py]: I also think we want to initial text datasets in a similar manner, where each dataset is wrapped in a class like in Task, which contains functions for sampling from the task, and evaluating, etc.

File: gato/training/trainer.py [hunk 4]
```diff
@@ -105,9 +108,13 @@ def train_iteration(self, num_steps, iter):
...
         logs['training/learning_rate'] = self.scheduler.get_lr()[0] # store LR at current step
-        
+
         # Build training batch
-        batch_dicts = self.sample_control_batch(self.args.batch_size)
+        # for simplicity sake - assume interleave training on control & text
+        if self.steps % 2 == 0:
+            batch_dicts = self.sample_control_batch(self.args.batch_size)
+        else:
+            batch_dicts = self.process_text_batch(next(iter(self.text_dataset)))
```
Comment (daniellawson9999) [inline gato/training/trainer.py]: the format of the dictionary currently returned by your process_text_batch is not compatible with input to forward() -> tokenize_input_dicts, as it expects each sequence to be a separate dictionary in a list. Some discussion on reworking this can be found here: https://github.com/ManifoldRG/gato-control/issues/5

File: gato/policy/gato_policy.py [hunk 5]
```diff
@@ -117,12 +117,12 @@ def __init__(
         self.separator_token = nn.Parameter(torch.zeros(embed_dim))
 
         # Tokenizers
-        self.text_tokenizer = None # e.g. SentencePiece
+        self.text_tokenizer = TextTokenizerGPT2()
```
Comment (daniellawson9999) [inline gato/policy/gato_policy.py]: make this configurable?

Comment (bhavul) [reply gato/policy/gato_policy.py]: Changing to AutoTokenizer since huggingface basically contains all major tokenizers neatly wrapped in an AutoTokenizer class. No point to duplicate.

File: gato/training/trainer.py (hunk 1 above)
Comment (bhavul) [reply gato/training/trainer.py]: Resolved in latest commit.

File: train.py (hunk 2 above)
Comment (bhavul) [reply train.py]: Agree, will do so in future commit.

File: gato/training/trainer.py (hunk 4 above)
Comment (bhavul) [reply gato/training/trainer.py]: Resolved in latest commit. Stuck to current format for now since #5 is still open.

File: train.py (hunk 2 above)
Comment (bhavul) [reply train.py]: done in latest commit including adding Task class

**Comment by bhavul** [comment]
Created #14 for this given our workflow currently always begins with an issue, and I'd forgotten creating one for this work. :)

File: gato/tasks/text_task.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        
+        if num_examples_to_test > len(self.text_dataset['test']):
+            print(f'num_examples_to_test chosen is more than test examples, so setting it to whole test dataset.')
+            num_examples_to_test = len(self.text_dataset['test'])
+        
+        for idx in range(num_examples_to_test):
+            text = self.text_dataset['test'][idx]['text']
+            target_text = self.text_dataset['test'][idx]['text']
+            
+            output_ids = model.predict_text(text, max_length=100, deterministic=deterministic)
```
Comment (daniellawson9999) [inline gato/tasks/text_task.py]: minor, maybe make this configurable, and another potential setting to make evaluation exhaustive, over all samples in the task

File: gato/tasks/text_task.py [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+from typing import List
+
+class TextTask(Task): 
+       
+    def __init__(self, task_type, dataset_names:List[str]):
+        super().__init__(task_type)
+        text_datasets_list = []
+        for text_dataset in dataset_names:
+            text_datasets_list.append(load_dataset(path='wikitext', name=text_dataset))
+        if len(text_datasets_list) == 1:
```
Comment (daniellawson9999) [inline gato/tasks/text_task.py:22]: The text task looks good so far, I am curious about the choice / what you think is the best way to handle having multiple  datasets. Are there speed benefits following the process here, of concatenating the datasets? If we had separate tasks, then we would also want to calculate the total tokens for each task / proportionally calculate how much of each batch comes from each task depending on # of tokens, but we don't have to worry about this if following your procedure. It seems like there is an edge case where the concatenation will not work if the columns are not named the same: https://huggingface.co/docs/datasets/process#concatenate

One thing that may be useful, is that if we have multiple datasets which are concatenated, is during evaluation, is to determine specific metrics associated w/ each separate dataset. E.g., want a separate perplexity score for wikitext vs the pile, not just the average between both. Potentially, after contenating, can maybe maintain start and end indices for each dataset, e.g. pile is 0 to 200mil, other dataset is (200mil + 1) to 400mil, so we can attribute which samples correspond to each task, separately aggregate their metrics. 

Another strategy is that during training, we just track the average, but after training finishes, you essentially load model, e.g. eval.py, just running over each of your tasks separately, where text_datasets={the specific dataset you want your eval metrics over}, but may be inconvenient

File: gato/tasks/text_task.py (hunk 6 above)
Comment (bhavul) [reply gato/tasks/text_task.py]: Wouldn't evaluation over all samples in the task be too much for text?

File: gato/tasks/text_task.py (hunk 7 above)
Comment (bhavul) [reply gato/tasks/text_task.py:22]: All valid points. Current strategy is simple.

- "there is an edge case where the concatenation will not work if the columns are not named the same" Yes but most good/large datasets do have them following the names. 
- "E.g., want a separate perplexity score for wikitext vs the pile, not just the average between both" - I'll raise a separate issue for this. I don't think this would be required anytime soon. We will switch to Pile once we are able to run on large scale data, and stick to only wikitext until then. For now, I imagine only one text dataset to be there. The first approach you mentioned here is better.

**Comment by bhavul** [comment]
The command used to test : 
```
python train.py --embed_dim=768 --layers=6 --heads=24 --training_steps=1000 --log_eval_freq=10 -
-warmup_steps=10 --batch_size=4 -k=240 --eval_episodes=10 --sequence_length=1024 --activation_fn=gelu --save_model --save_mode=checkpoint --text_prop=1.0 --cpu 
```

Pasting some minor logs, we do see perplexity and loss decreasing...
```
Trainable Parameters: 124.451715M
================================================================================
Iteration 0
training/learning_rate: 9.001e-05
time/training: 11.375323057174683
evaluation/text/loss: 6.627391033172607
evaluation/text/perplexity: 755.5083618164062
time/total: 265.7239911556244
time/evaluation: 254.34865498542786
training/train_loss_mean: 10.707273292541505
training/train_loss_std: 0.40707113294316866
================================================================================
Iteration 1
training/learning_rate: 9.998164867726852e-05
time/training: 7.935048818588257
evaluation/text/loss: 6.625466632843017
evaluation/text/perplexity: 754.05615234375
time/total: 345.6292951107025
time/evaluation: 71.22406029701233
training/train_loss_mean: 9.438685131072997
training/train_loss_std: 0.418305974439971
================================================================================
Iteration 2
training/learning_rate: 9.991823122200031e-05
time/training: 11.019861936569214
evaluation/text/loss: 6.604997024536133
evaluation/text/perplexity: 738.7777709960938
time/total: 446.92891120910645
time/evaluation: 89.53751707077026
training/train_loss_mean: 8.593882703781128
training/train_loss_std: 0.43145140181362285
================================================================================
Iteration 3
training/learning_rate: 9.980958491544698e-05
time/training: 10.03787112236023
evaluation/text/loss: 6.5821265029907225
evaluation/text/perplexity: 722.0732421875
time/total: 566.3537652492523
time/evaluation: 108.6496319770813
training/train_loss_mean: 8.053662157058715
training/train_loss_std: 0.23972693794669167
================================================================================
Iteration 4
training/learning_rate: 9.965581915522964e-05
time/training: 12.212831974029541
evaluation/text/loss: 6.593824853897095
evaluation/text/perplexity: 730.5698852539062
time/total: 684.9295170307159
time/evaluation: 105.59774208068848
training/train_loss_mean: 7.210125470161438
training/train_loss_std: 1.442274285697434
```

The PR can be merged now. I will add details of longer runs separately once I get it working on Collab.
cc @daniellawson9999 @harshsikka

File: gato/tasks/text_task.py (hunk 7 above)
Comment (bhavul) [reply gato/tasks/text_task.py:22]: Added #22.

**Comment by daniellawson9999** [comment]
Awesome, currently reviewing, found that some errors when running on GPU, specifically, some tensors still are on CPU. Adding modifications, simply adding .to(device=self.device) in a couple places, like here:

https://github.com/ManifoldRG/NEKO/blob/b3fe09549855c99f3e96ff6943a20fd959f33968/gato/policy/gato_policy.py#L258

File: gato/policy/gato_policy.py [hunk 8]
```diff
@@ -89,6 +91,7 @@ def __init__(
                 config=config,
             )
             embed_dim = config.n_embd
+            self.embed_token = self.transformer.wte
```
Comment (daniellawson9999) [inline gato/policy/gato_policy.py]: I want to modify this line. The size of self.transformer.wte should be equal to text_tokens, but you want to expand your nn.Embeddding to go up to vocab_size. I can work on modifying this, following from here https://github.com/machelreid/can-wikipedia-help-offline-rl/blob/e1060f0e03a22e3c05fb70cbad05dfc840ce8afc/code/decision_transformer/models/decision_transformer.py#L78 (the code here is expanding the embedding dict for positional embeddings not token embeddings but same logic applies)

File: gato/policy/gato_policy.py [hunk 9]
```diff
@@ -89,6 +82,11 @@ def __init__(
                 config=config,
             )
             embed_dim = config.n_embd
+            #self.embed_token = self.transformer.wte
+            assert self.transformer.wte.weight.shape[0] == self.text_tokens, "pretrained token/expected mimsatch" # potentially make text_tokens dynamic
+            # expand embedding dictionary up to vocab_size
+            self.embed_token = nn.Embedding(self.vocab_size, embed_dim)
+            self.embed_token.weight.data[:self.text_tokens] = self.transformer.wte.weight.data
```
Comment (daniellawson9999) [inline gato/policy/gato_policy.py:95]: I updated this to expand the pretrained wte

Comment (bhavul) [reply gato/policy/gato_policy.py:95]: Perhaps I didn't encounter this since I was always using text_prop as 1.0 and doing an isolated testing. This makes a lot of sense. Thanks for the addition!

File: gato/policy/gato_policy.py (hunk 8 above)
Comment (bhavul) [reply gato/policy/gato_policy.py]: Go for it! Thanks for the support. :)

File: gato/policy/gato_policy.py [hunk 10]
```diff
@@ -48,9 +51,7 @@ def __init__(
 
         self.context_len = context_len
         self.pad_seq = pad_seq
-        # this is a dummy value as this implementation does not yet handle language IO
-        #self.text_tokens = 32000 # SentencePiece vocab size
-        self.text_tokens = 1
+        self.text_tokens = 50257 # gpt2
```
Comment (daniellawson9999) [inline gato/policy/gato_policy.py]: make this configurable in the future e.g. decide this after you load GPT2 and then infer this

**Comment by bhavul** [comment]
Just wanted to add an update that I'm revamping the way text tokenization is happening in the codebase to follow some industry practices and hence this PR merge is delayed a bit.

**Comment by bhavul** [comment]
**Couple of runs with just text** :
1. Only Text from scratch : https://wandb.ai/bhavul/gato-control/runs/zckzf7tr/overview?workspace=user-bhavul 
2. Only text with pretrained_lm GPT2 : https://wandb.ai/bhavul/gato-control/runs/arv80xmv/overview?workspace=user-bhavul

Both contain the CLI command used for them. Make sure to be in the environment.

Running with control datasets with 0.5 weightage to control and 0.5 to text (**WIP**). Note that you can not train with control datasets on Mac machines with Apple Silicone (M1 / M2) architectures, since the underlying mujoco datasets interface requires gcc9.x which isn't built for macs with apple silicon.

**Comment by bhavul** [comment]
Latest results shared on discord.

I got decent perplexity and eval text loss scores in just 16 steps of training with this command (using pretrained GPT2) : 
```
python train.py --embed_dim=768 --layers=6 --heads=24 --training_steps=1000 --log_eval_freq=10 --warmup_steps=20 --batch_size=16 --sequence_length=1024 --eval_episodes=10 --activation_fn=gelu --save_model --save_mode=checkpoint --text_prop=1.0 --eval_text_log_examples --text_datasets=wikitext-2-v1 --text_datasets_paths=wikitext --use_wandb --pretrained_lm=gpt2 --disable_cosine_decay
```

It seemed to not be overfitting in this case. Both training loss and eval loss were decreasing, although I couldn't run for very long. It did run CUDA OOM after a certain point - so we may need beefier GPUs, or reduce num of layers, embed_dim, etc.

I was also able to find a command that quickly reduces training loss at the expense of overfitting quickly (and having bad eval loss, at least in early 4-5 steps). This has lesser number of layers, and does not use pretrained GPT2. 

```
python train.py --embed_dim=512 --layers=4 --heads=8 --training_steps=100000 --log_eval_freq=500 --warmup_steps=10000 --batch_size=16 --sequence_length=1024 --eval_episodes=10 --activation_fn=gelu --disable_cosine_decay --text_prop=1.0 --eval_text_log_examples --flash --gradient_accumulation_steps=1 --text_datasets=wikitext-2-v1 --text_datasets_paths=wikitext --use_wandb
```

Overall, the code does work however now. It's a point of finding good hyperparameters, or train it for long so we're able to see some good results. 

I'm merging the code now.

**Comment by bhavul** [comment]
Here's a collab notebook as well if someone wishes to test things themselves : https://colab.research.google.com/drive/12kDwrQ3Sfm3zAtYXdqFdu0bYisIfBawm?usp=sharing
